pydantic/main.py

```diff
@@ -584,34 +602,45 @@ def _iter(
         allowed_keys: Optional['SetStr'] = None,
         include: Union['SetIntStr', 'DictIntStrAny'] = None,
         exclude: Union['SetIntStr', 'DictIntStrAny'] = None,
-        skip_defaults: bool = False,
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
     ) -> 'TupleGenerator':
 
         value_exclude = ValueItems(self, exclude) if exclude else None
         value_include = ValueItems(self, include) if include else None
 
+        if exclude_defaults:
+            if allowed_keys is None:
+                allowed_keys = set(self.__fields__)
+            for k in allowed_keys:
+                if self.__dict__[k] == self.__fields__[k].default:
+                    allowed_keys.discard(k)
```

Comment:
@samuelcolvin does the above logic seem reasonable to you for how to perform this? I initially put it in the `for k, v in self.__dict__.items()` loop, but this approach results in fewer checks when `exclude_defaults` is `False` (which I think will be the most common scenario).

Also, this feels like another small nudge toward storing the field defaults in their own dict for fast lookup. Given https://github.com/samuelcolvin/pydantic/pull/898, do you think it could make sense to add that?

I'm fine either way.

Reply:
this looks like it's duplicated below - presumably you meant that to demonstrate both approaches.

This route looks preferable to me.

---

I agree with you on `self.__defaults__` especially if the values are somehow shared between `self.__fields__[k].default` and `self.__defaults__[k]`. 

It would be as simple as adding:

```py
'__defaults__': {n: f.default for n, f in fields.items()},
```
to `new_namespace` in `ModelMetaclass`

Reply:
I didn't actually mean it in both cases 🤦‍♂️ hard to parse the diffs with all the renames. Thanks for all this feedback.

Reply:
@samuelcolvin how should this be handled if the field is required? Should I just put the non-required defaults in `__defaults__`? That's what we'd want for #898 too.

Reply:
yes, `if not f.required` makes sense.

I just tried this on #898 and pycharm complained that `__defaults__` is a special name, see [here](https://docs.python.org/3/reference/datamodel.html). 

Do you think it's okay to call it that? I'm not at all clear whether it might break something.

Reply:
I implemented this where `__defaults__` only holds the values for non-required fields. I think it's basically optimally fast when checking to exclude defaults -- you'd have to check them all, and they are cached. This should also simplify the logic for #898.

Reply:
Good catch on the naming -- let's call it `__field_defaults__`? I'll rename it to that, but let me know if you'd prefer something else.